apps/context.go

```diff
@@ -75,8 +76,12 @@ type ExpandedContext struct {
 	User *model.User `json:"user,omitempty"`
 }
 
-// Paths for OAuth endpoint that the App needs to set up remote (3rd party) OAuth
-const (
-	PathOAuthRedirect = "/oauth2/remote/redirect"
-	PathOAuthComplete = "/oauth2/remote/complete"
-)
+type OAuth2Context struct {
+	// Expanded with "oauth2_app". Config must be previously stored with
+	// mmclient.StoreOAuth2App
+	OAuth2App
+	RedirectURL string `json:"redirect_url,omitempty"`
```

Comment:
Just to be clear. The redirect URL is the place I have to send someone to start the OAuth flow, right?
I post the link to the redirect URL, this goes to a mattermost endpoint that ask the app for the OAuth link to start the flow (called also redirect URL), and while giving that information, I store the state.

The complete URL is the URL where I should redirect the user after the flow is finished, right? When I create the link to start the flow, I should do it with the complete URL as the redirect at the end of the flow. When that endpoint is hit, the proxy will forward to the app the query values, and the app will check the state, create the token, and store the user.

The naming seems a bit weird right now, because there are several "redirect" elements here and there. Any idea how to make it more clear? I feel the redirectURL here in particular could change to something like "StartURL", since what is doing is start the flow.

Reply:
I agree. I wish the OAuth2 package (and spec?) used the client-centric term "Complete" for the "final redirect"

`start`? `connect`? and `complete`

Reply:
I like the idea of `start` and `connect`, but I am struggling with `complete`.

`start` starts the process on mattermost side
`connect` is the url to the OAuth provider. Mattermost will redirect from start to the connect url.
`complete` is the url to which the OAuth provider will redirect after finishing. That is a good fit for complete naming, but it is also the `redirect url` as per OAuth spec. 0/5 on which direction to go.

Reply:
Current flow:
1. "connect link" HTTP `/apps/appid/oauth2/remote/redirect` (A)
2. served by PP, invokes `GetOAuth2RedirectURL` call (B) which returns a "redirect URL" (C), issues a 302 there
3. upon completion, the browser is redirected to `/apps/appid/oath2/remote/complete` (D)
4. state verified, served by PP, invokes `OnOAuth2Complete` (E)
5. app stores the token as "User" (cc @hanzei)

Proposed naming:
- A: `/apps/appid/oauth2/remote/connect`
- B: `"oauth2_connect"`
- C: connect URL
- D: no change
- E: `"oauth2_complete"`